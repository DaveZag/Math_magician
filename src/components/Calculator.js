import React, { PureComponent } from 'react';

import Button from './Button';
import './calculator.css';
import calculate from '../logic/calculate';

const signArray = [
  'AC',
  '+/-',
  '%',
  'รท',
  '7',
  '8',
  '9',
  'x',
  '4',
  '5',
  '6',
  '-',
  '1',
  '2',
  '3',
  '+',
  '0',
  '.',
  '=',
];

class Calculator extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      total: '0',
      next: null,
      operation: null,
    };
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick = (e) => {
    const prevState = this.state;
    this.setState(calculate(prevState, e.target.innerText));
  };

  render() {
    const { total, next, operation } = this.state;
    const checkValues = () => {
      if (total || operation || next) {
        return `${total || ''} ${operation || ''} ${next || ''} `;
      }
      return '0';
    };

    return (
      <div className="calc-container">
        <input className="output" value={checkValues()} disabled />

        <div className="grid">
          {signArray.map((sign) => (
            <Button
              key={`unique-${sign}`}
              text={sign}
              onClick={this.handleClick}
            />
          ))}
        </div>
      </div>
    );
  }
}

export default Calculator;
