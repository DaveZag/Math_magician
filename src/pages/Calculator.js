import React, { useState } from 'react';

import Button from '../components/Button';
import './calculator.css';
import calculate from '../logic/calculate';

const Calculator = () => {
  const signArray = [
    'AC',
    '+/-',
    '%',
    'รท',
    '7',
    '8',
    '9',
    'x',
    '4',
    '5',
    '6',
    '-',
    '1',
    '2',
    '3',
    '+',
    '0',
    '.',
    '=',
  ];

  const [dataObj, setDataObj] = useState({
    total: '0',
    next: null,
    operation: null,
  });

  const handleClick = (e) => {
    setDataObj(calculate(dataObj, e.target.textContent));
  };

  const { total, next, operation } = dataObj;
  const checkValues = () => {
    if (total || operation || next) {
      return `${total || ''} ${operation || ''} ${next || ''} `;
    }
    return '0';
  };

  return (
    <section className="calculator-section flex flex-col">
      <header>
        <h2 className="welcome-txt">Let&apos;s do some math!</h2>
      </header>

      <div className="calculator flex jc-c ai-c" data-testid="calculator">
        <div className="calc-container">
          <input
            className="output"
            data-testid="screen"
            value={checkValues()}
            disabled
          />

          <div className="grid">
            {signArray.map((sign) => (
              <Button
                key={`unique-${sign}`}
                text={sign}
                onClick={handleClick}
              />
            ))}
          </div>
        </div>
      </div>
    </section>
  );
};

export default Calculator;
