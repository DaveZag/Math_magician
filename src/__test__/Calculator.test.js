import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Calculator from '../pages/Calculator';

describe('Tests for the calculator component and functions it calls', () => {
  it('Should render the calculator page correctly', () => {
    const calcPage = render(<Calculator />);
    expect(calcPage).toMatchSnapshot();
  });

  it('Should render the calculator', () => {
    render(<Calculator />);
    const calculator = screen.getByTestId('calculator');
    expect(calculator).toMatchSnapshot();
  });

  it("should render the button's text on the calculator screen when clicked", () => {
    render(<Calculator />);
    const button = screen.getByRole('button', { name: '7' });
    const calcScreen = screen.getByTestId('screen');

    fireEvent.click(button);
    expect(calcScreen.value.trim()).toBe('7');
  });

  it('should render the number to be added', () => {
    render(<Calculator />);
    const button8 = screen.getByRole('button', { name: '5' });
    const buttonAdd = screen.getByRole('button', { name: '+' });
    const button9 = screen.getByRole('button', { name: '9' });

    const calcScreen = screen.getByTestId('screen');

    fireEvent.click(button8);
    fireEvent.click(buttonAdd);
    fireEvent.click(button9);
    expect(calcScreen.value.trim()).toBe('5 + 9');
  });

  it('should correctly add the inputted numbers', () => {
    render(<Calculator />);
    const button8 = screen.getByRole('button', { name: '8' });
    const buttonAdd = screen.getByRole('button', { name: '+' });
    const button9 = screen.getByRole('button', { name: '9' });
    const buttonEquals = screen.getByRole('button', { name: '=' });

    const calcScreen = screen.getByTestId('screen');

    fireEvent.click(button8);
    fireEvent.click(buttonAdd);
    fireEvent.click(button9);
    fireEvent.click(buttonEquals);
    expect(calcScreen.value.trim()).toBe('17');
  });

  it('should correctly multiply the inputted numbers', () => {
    render(<Calculator />);
    const button5 = screen.getByRole('button', { name: '5' });
    const buttonMultiply = screen.getByRole('button', { name: 'x' });
    const button8 = screen.getByRole('button', { name: '8' });
    const buttonEquals = screen.getByRole('button', { name: '=' });

    const calcScreen = screen.getByTestId('screen');

    fireEvent.click(button5);
    fireEvent.click(buttonMultiply);
    fireEvent.click(button8);
    fireEvent.click(buttonEquals);
    expect(calcScreen.value.trim()).toBe('40');
  });
});
