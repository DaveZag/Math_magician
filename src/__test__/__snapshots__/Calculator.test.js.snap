// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the calculator component and functions it calls Should render the calculator 1`] = `
<div
  class="calculator flex jc-c ai-c"
  data-testid="calculator"
>
  <div
    class="calc-container"
  >
    <input
      class="output"
      data-testid="screen"
      disabled=""
      value="0   "
    />
    <div
      class="grid"
    >
      <button
        class="button-cont"
        type="button"
      >
        AC
      </button>
      <button
        class="button-cont"
        type="button"
      >
        +/-
      </button>
      <button
        class="button-cont"
        type="button"
      >
        %
      </button>
      <button
        class="button-cont"
        type="button"
      >
        รท
      </button>
      <button
        class="button-cont"
        type="button"
      >
        7
      </button>
      <button
        class="button-cont"
        type="button"
      >
        8
      </button>
      <button
        class="button-cont"
        type="button"
      >
        9
      </button>
      <button
        class="button-cont"
        type="button"
      >
        x
      </button>
      <button
        class="button-cont"
        type="button"
      >
        4
      </button>
      <button
        class="button-cont"
        type="button"
      >
        5
      </button>
      <button
        class="button-cont"
        type="button"
      >
        6
      </button>
      <button
        class="button-cont"
        type="button"
      >
        -
      </button>
      <button
        class="button-cont"
        type="button"
      >
        1
      </button>
      <button
        class="button-cont"
        type="button"
      >
        2
      </button>
      <button
        class="button-cont"
        type="button"
      >
        3
      </button>
      <button
        class="button-cont"
        type="button"
      >
        +
      </button>
      <button
        class="button-cont"
        type="button"
      >
        0
      </button>
      <button
        class="button-cont"
        type="button"
      >
        .
      </button>
      <button
        class="button-cont"
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`Tests for the calculator component and functions it calls Should render the calculator page correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="calculator-section flex flex-col"
      >
        <header>
          <h2
            class="welcome-txt"
          >
            Let's do some math!
          </h2>
        </header>
        <div
          class="calculator flex jc-c ai-c"
          data-testid="calculator"
        >
          <div
            class="calc-container"
          >
            <input
              class="output"
              data-testid="screen"
              disabled=""
              value="0   "
            />
            <div
              class="grid"
            >
              <button
                class="button-cont"
                type="button"
              >
                AC
              </button>
              <button
                class="button-cont"
                type="button"
              >
                +/-
              </button>
              <button
                class="button-cont"
                type="button"
              >
                %
              </button>
              <button
                class="button-cont"
                type="button"
              >
                รท
              </button>
              <button
                class="button-cont"
                type="button"
              >
                7
              </button>
              <button
                class="button-cont"
                type="button"
              >
                8
              </button>
              <button
                class="button-cont"
                type="button"
              >
                9
              </button>
              <button
                class="button-cont"
                type="button"
              >
                x
              </button>
              <button
                class="button-cont"
                type="button"
              >
                4
              </button>
              <button
                class="button-cont"
                type="button"
              >
                5
              </button>
              <button
                class="button-cont"
                type="button"
              >
                6
              </button>
              <button
                class="button-cont"
                type="button"
              >
                -
              </button>
              <button
                class="button-cont"
                type="button"
              >
                1
              </button>
              <button
                class="button-cont"
                type="button"
              >
                2
              </button>
              <button
                class="button-cont"
                type="button"
              >
                3
              </button>
              <button
                class="button-cont"
                type="button"
              >
                +
              </button>
              <button
                class="button-cont"
                type="button"
              >
                0
              </button>
              <button
                class="button-cont"
                type="button"
              >
                .
              </button>
              <button
                class="button-cont"
                type="button"
              >
                =
              </button>
            </div>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="calculator-section flex flex-col"
    >
      <header>
        <h2
          class="welcome-txt"
        >
          Let's do some math!
        </h2>
      </header>
      <div
        class="calculator flex jc-c ai-c"
        data-testid="calculator"
      >
        <div
          class="calc-container"
        >
          <input
            class="output"
            data-testid="screen"
            disabled=""
            value="0   "
          />
          <div
            class="grid"
          >
            <button
              class="button-cont"
              type="button"
            >
              AC
            </button>
            <button
              class="button-cont"
              type="button"
            >
              +/-
            </button>
            <button
              class="button-cont"
              type="button"
            >
              %
            </button>
            <button
              class="button-cont"
              type="button"
            >
              รท
            </button>
            <button
              class="button-cont"
              type="button"
            >
              7
            </button>
            <button
              class="button-cont"
              type="button"
            >
              8
            </button>
            <button
              class="button-cont"
              type="button"
            >
              9
            </button>
            <button
              class="button-cont"
              type="button"
            >
              x
            </button>
            <button
              class="button-cont"
              type="button"
            >
              4
            </button>
            <button
              class="button-cont"
              type="button"
            >
              5
            </button>
            <button
              class="button-cont"
              type="button"
            >
              6
            </button>
            <button
              class="button-cont"
              type="button"
            >
              -
            </button>
            <button
              class="button-cont"
              type="button"
            >
              1
            </button>
            <button
              class="button-cont"
              type="button"
            >
              2
            </button>
            <button
              class="button-cont"
              type="button"
            >
              3
            </button>
            <button
              class="button-cont"
              type="button"
            >
              +
            </button>
            <button
              class="button-cont"
              type="button"
            >
              0
            </button>
            <button
              class="button-cont"
              type="button"
            >
              .
            </button>
            <button
              class="button-cont"
              type="button"
            >
              =
            </button>
          </div>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
